src/Microsoft.Azure.WebJobs.Host/Loggers/Logger/ApplicationInsights/ApplicationInsightsLoggerProvider.cs

```diff
@@ -13,22 +13,47 @@ internal class ApplicationInsightsLoggerProvider : ILoggerProvider
         private readonly TelemetryClient _client;
         private readonly Func<string, LogLevel, bool> _filter;
 
-        public ApplicationInsightsLoggerProvider(string instrumentationKey, Func<string, LogLevel, bool> filter,
-            ITelemetryClientFactory clientFactory, SamplingPercentageEstimatorSettings samplingSettings)
+        private ITelemetryClientFactory _clientFactory;
+        private bool _disposed;
+
+        public ApplicationInsightsLoggerProvider(Func<string, LogLevel, bool> filter, ITelemetryClientFactory clientFactory)
         {
-            if (instrumentationKey == null)
+            if (clientFactory == null)
             {
-                throw new ArgumentNullException(nameof(instrumentationKey));
+                throw new ArgumentNullException(nameof(clientFactory));
             }
 
-            _client = clientFactory.Create(instrumentationKey, samplingSettings);
+            _clientFactory = clientFactory;
+            _client = _clientFactory.Create();
             _filter = filter;
         }
 
         public ILogger CreateLogger(string categoryName) => new ApplicationInsightsLogger(_client, categoryName, _filter);
 
         public void Dispose()
         {
+            if (!_disposed)
+            {
+                _clientFactory.Dispose();
+
+                if (_client != null)
+                {
+                    try
+                    {
+                        _client.Flush();
+
+                        // This sleep isn't ideal, but the flush is async so it's the best we have right now. This is
+                        // being tracked at https://github.com/Microsoft/ApplicationInsights-dotnet/issues/407
+                        Thread.Sleep(2000);
```

Comment:
Well, in testing E2E I see data loss without it. The TelemetryChannel we're using caches to the local file system so that anything it missed should eventually be sent. However, it'd be nicer if we didn't have to restart the process to pick it back up.

The Channel fires up a separate thread to send the data and we never see it. The suggestion to guarantee delivery was to wait a full minute, but that's not really reasonable in a shutdown path like this. In my experience, I was seeing my data come through with a two-second wait, but a one-second wait was not enough. I'd also be willing to go longer, but it'll apply to every host shutdown and still not even be guaranteed.

As you can see from the linked issue, others have this problem as well. As soon as there's a fix available, we can swap to use it.